eZ/Publish/Core/Repository/UserService.php

```diff
@@ -570,9 +572,38 @@ public function loadUserByCredentials($login, $password, array $prioritizedLangu
             throw new NotFoundException('user', $login);
         }
 
+        $this->updateDeprecatedPasswordHash($login, $password, $spiUser);
+
         return $this->buildDomainUserObject($spiUser, null, $prioritizedLanguages);
     }
 
+    /**
+     * Update password hash to the type configured for the service, if the hash is of a deprecated type.
+     *
+     * @param string $login User login
+     * @param string $password User password
+     * @param \eZ\Publish\SPI\Persistence\User $spiUser
+     */
+    private function updateDeprecatedPasswordHash($login, $password, SPIUser $spiUser)
+    {
+        if ($spiUser->hashAlgorithm !== $this->settings['hashType'] &&
+            in_array(
```

Comment:
is this logic in legacy? I got the impression legacy will start migrating over hashes as soon config has changed?

Reply:
Legacy updates non-default hash (deprecated or not) to current default if setting `UpdateHash` is true (default). So you can disable it.

This was a conservative approach. We could skip the deprecated check and rename it to `updatePasswordHash()`. (Don't want to add another setting for this.)

Reply:
+1